### ISSUE TITLE
"Too many" tries in Article test causes a crash.

### ISSUE THREAD

**Issue by jdickey** [jdickey/article_fixture_gen#13]
Consider the `Article` tests that generate a number of instances of `Article` to evaluate the number of paragraphs within the generated body markup, whether configured for a [range of allowable counts](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/test/article_fixture_gen/data/article_test.rb#L68-L82) or a [fixed count](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/test/article_fixture_gen/data/article_test.rb#L84-L94).

On our development system, raising the `:try_count` control variable above its current 10 will generate an increasing number of `nil`-reference failures. Changing the setting on [line 71](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/test/article_fixture_gen/data/article_test.rb#L71) from the current `10` to `100` and running just the [allowable-range-count tests](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/test/article_fixture_gen/data/article_test.rb#L68-L82) produces output such as the following:

```
$ ruby test/article_fixture_gen/data/article_test.rb

# Running tests with run options --seed 36070:

.E
Error:
ArticleFixtureGen::Data::Article::has a #to_s method that::when an allowable range of paragraph counts is specified#test_0001_generates a minimum of :para_count_min paragraphs in the content:
NoMethodError: undefined method `join' for nil:NilClass
Did you mean?  JSON
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/split_text_at_target_word.rb:41:in `tail'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/split_text_at_target_word.rb:47:in `tail_index'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/split_text_at_target_word.rb:37:in `head'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/split_text_at_target_word.rb:16:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/split_text_at_target_word.rb:12:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/replace_child_nodes_using_new_child.rb:45:in `nodes'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/replace_child_nodes_using_new_child.rb:16:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/replace_child_nodes_using_new_child.rb:12:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:56:in `add_smtp'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:52:in `block in add_entries'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:52:in `each'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:52:in `add_entries'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:34:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/smtp_decorated_markup.rb:29:in `call'
    /Users/jeffdickey/src/rails/meldd/article_fixture_gen/lib/article_fixture_gen/data/article.rb:24:in `to_s'
    test/article_fixture_gen/data/article_test.rb:10:in `count_paras_in'
    test/article_fixture_gen/data/article_test.rb:21:in `block in count_paras'
    test/article_fixture_gen/data/article_test.rb:21:in `map'
    test/article_fixture_gen/data/article_test.rb:21:in `count_paras'
    test/article_fixture_gen/data/article_test.rb:76:in `block (4 levels) in <main>'


Finished tests in 4.169930s, 0.4796 tests/s, 0.2398 assertions/s.


2 tests, 1 assertions, 0 failures, 1 errors, 0 skips
Coverage report generated for Unit Tests to /Users/jeffdickey/src/rails/meldd/article_fixture_gen/coverage. 299 / 299 LOC (100.0%) covered.
$
```

Two possibilities for the defect responsible for the fault come to mind:

1. The [`#tail_words` method](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/lib/article_fixture_gen/data/split_text_at_target_word.rb#L50-L52) is returning `nil` because the [passed_in `word_index` value](https://github.com/jdickey/article_fixture_gen/blob/18a3a2d/lib/article_fixture_gen/data/split_text_at_target_word.rb#L24) is out of range in certain cases that occur rarely enough to be undetectable in 50 test runs when the `try_count` setting is 10 or lower, and increasingly common when raised above 10;
2. It is *possible* that there is a defect in the MRI memory management that we exercise when repeatedly slicing-and-dicing the same base markup string.

> I'll take Door Number One, Monty!

**Comment by jdickey** [comment]
There *is* an obvious fix for this…that apparently masks a deeper bug; how can `#tail_words` possibly return `nil` based on (supposedly) trustworthy inputs?

*Pro tip:* To repeatedly run a test using the same seed with MRI Ruby 2.*x* (which no longer directly supports the `--seed` option to pass to MiniTest), use something like the following:

```
$ ruby -e 'require "./test/article_fixture_gen/data/article_test.rb"' -- --seed 36070
```

That command line, when run against Commit 18a3a2d, will produce the output quoted in the opening comment.

**Comment by mitpaladin** [comment]
Some of the text are not in `@nodes` but outside in `@value`, is that supposed to be the case?  

here's the problematic instance with seed 18339, captured with `binding.pry if tail_words.nil?` in `split_text_at_target_word.rb`
```
[1] pry(#<ArticleFixtureGen::Data::SplitTextAtTargetWord>)> el
=> #<Ox::Element:0x007fc09405c268
 @nodes=
  ["Asperiores laborum odio ea omnis consequatur optio maxime. Qui dicta illum debitis earum dolorum nostrum reiciendis. Nulla quis nam error harum nemo et labore vero. Excepturi quas assumenda exercitationem consequatur quidem omnis corrupti voluptatem. ",
   #<Ox::Element:0x007fc094057858
    @attributes={:href=>"#ea", :title=>"Asperiores laboriosam"},
    @nodes=["Rerum consequuntur"],
    @value="a">,
   #<Ox::Element:0x007fc094057650 @nodes=["doloremque id"], @value="code">,
   #<Ox::Element:0x007fc0940573d0 @nodes=["Repudiandae in"], @value="strong">,
   ". Id officiis eius quod assumenda quia consequatur ut ",
   #<Ox::Element:0x007fc095896900
    @attributes={:href=>"", :id=>"single-mtp-43"},
    @nodes=[],
    @value="a">,
   "aut. ",
   #<Ox::Element:0x007fc094056fe8
    @nodes=
     ["Et libero itaque asperiores quo omnis. Esse doloremque aspernatur dolorum optio sed porro. Adipisci et ullam at soluta. In consequatur velit repellat et quae et et voluptatem. Cupiditate quibusdam accusamus maxime aut vitae facere."],
    @value="em">],
 @value="p">
[2] pry(#<ArticleFixtureGen::Data::SplitTextAtTargetWord>)> target_entry
=> #<ArticleFixtureGen::Data::WordEntry dom_position=[6, 4] word="quas" word_index=13>
```

at the Build SMTP Entry List level the parent node was 
```
   #<Ox::Element:0x007fd6320f4640
    @nodes=
     ["Asperiores laborum odio ea omnis consequatur optio maxime. Qui dicta illu
m debitis earum dolorum nostrum reiciendis. Nulla quis nam error harum nemo et l
abore vero. Excepturi quas assumenda exercitationem consequatur quidem omnis cor
rupti voluptatem. ",
      #<Ox::Element:0x007fd6320f44b0
       @attributes={:href=>"#ea", :title=>"Asperiores laboriosam"},
       @nodes=["Rerum consequuntur"],
       @value="a">,
      #<Ox::Element:0x007fd6320f4208 @nodes=["doloremque id"], @value="code">,
      #<Ox::Element:0x007fd6320eff28
       @nodes=["Repudiandae in"],
       @value="strong">,
      ". Id officiis eius quod assumenda quia consequatur ut aut. Vero quia itaque quas dicta facilis. Totam est praesentium aut suscipit. Esse et voluptatem perspiciatis rerum voluptatum blanditiis sequi eaque. Veritatis quibusdam alias culpa perferendis voluptas. Voluptas esse excepturi quis veritatis expedita. Tempore magni occaecati dolor veniam sit. Molestiae ut quas asperiores sunt. Cum autem et consequatur rem reprehenderit nihil aut. ",
      #<Ox::Element:0x007fd6320efde8
       @nodes=
        ["Et libero itaque asperiores quo omnis. Esse doloremque aspernatur dolorum optio sed porro. Adipisci et ullam at soluta. In consequatur velit repellat et quae et et voluptatem. Cupiditate quibusdam accusamus maxime aut vitae facere."],
       @value="em">],
    @value="p">,
```
so the insertion of single-mtp-43 lost "aut. Vero quia itaque quas dicta facilis. Totam est praesentium aut suscipit. Esse et voluptatem perspiciatis rerum voluptatum blanditiis sequi eaque. Veritatis quibusdam alias culpa perferendis voluptas. Voluptas esse excepturi quis veritatis expedita. Tempore magni occaecati dolor veniam sit. Molestiae ut quas asperiores sunt. Cum autem et consequatur rem reprehenderit nihil" in `Ox::Element:0x007fd6320eff28` while the subsequent insertion is looking for the "quas"

**Comment by mitpaladin** [comment]
Expected behavior from insertion of   `single-mtp-43' should have been:
```
#<Ox::Element:0x007fc0940573d0 @nodes=["Repudiandae in"], @value="strong">,
   ". Id officiis eius quod assumenda quia consequatur ut aut. Vero quia itaque quas dicta facilis. Totam est praesentium aut suscipit. Esse et voluptatem perspiciatis rerum voluptatum blanditiis sequi eaque. Veritatis quibusdam alias culpa perferendis voluptas. Voluptas esse excepturi quis veritatis expedita. Tempore magni occaecati dolor veniam sit. Molestiae ut quas asperiores sunt. Cum autem et consequatur rem reprehenderit nihil",
   #<Ox::Element:0x007fc095896900
    @attributes={:href=>"", :id=>"single-mtp-43"},
    @nodes=[],
    @value="a">,
   "aut. ",
'''
correct?

**Comment by jdickey** [comment]
What you're showing is

* a top-level `p` element which has multiple child nodes:
  * A text string ("Asperiores labour audio ea...")
  * An `<a>` element, with the text `Rerum consequuntur` and `href` and `title` attributes;
  * A `<code>` element with the text `doloremque id`;
  * A `<strong>` element with text `. Id offices emus quod assumenda...`; and
  * An `<em>` element with text `Et libber itaque asperiores quo omens. ...".

*The text is boilerplate*; that's what Lorem Ipsum *means*, yes?

Where was the sMTP supposed to be inserted? What leads you to that conclusion? There's not enough information here for me to figure out what you're seeing in detail.

**Comment by mitpaladin** [comment]
So the `single-mtp-43` split bug seems to be due to the fact that the target_entry `aut.` is repeated within the element

```
[5] pry(#<ArticleFixtureGen::Data::SplitTextAtTargetWord>)> original_text
=> ". Id officiis eius quod assumenda quia consequatur ut aut. Vero quia itaque quas dicta facilis. Totam est praesentium aut suscipit. Esse et voluptatem perspiciatis rerum voluptatum blanditiis sequi eaque. Veritatis quibusdam alias culpa perferendis voluptas. Voluptas esse excepturi quis veritatis expedita. Tempore magni occaecati dolor veniam sit. Molestiae ut quas asperiores sunt. Cum autem et consequatur rem reprehenderit nihil aut. "
```

head returns the text prior to the first `aut.` while tail returns the text after the last `aut.`
```
[2] pry(#<ArticleFixtureGen::Data::SplitTextAtTargetWord>)> head
=> ". Id officiis eius quod assumenda quia consequatur ut "
[3] pry(#<ArticleFixtureGen::Data::SplitTextAtTargetWord>)> tail
=> "aut. "
```

I'm sure the fix needs to be the calculation of `tail_index`,
```
      def tail_index
        original_text.index(tail)
      end
```
as the String index is the wrong operation here, but rather needs to directly use the `word_index`

**Comment by jdickey** [comment]
Closed by Commit a74828a.
